File: backend/src/sensei/models/a3.py
"""
A3 Problem Solving models.

Implements:
- A3: A3 problem-solving document
- A3Section: Individual sections of an A3
"""

from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING
from uuid import UUID

from sqlalchemy import (
    Boolean,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import JSONB, UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from sensei.models.base import AuditMixin, Base, SoftDeleteMixin, TimestampMixin

if TYPE_CHECKING:
    from sensei.models.user import User


class A3Type(str, Enum):
    """Type of A3 document."""
    
    PROBLEM_SOLVING = "problem_solving"
    PROPOSAL = "proposal"
    STATUS_REPORT = "status_report"
    STRATEGY = "strategy"


class A3Status(str, Enum):
    """Status of A3 document."""
    
    DRAFT = "draft"
    IN_PROGRESS = "in_progress"
    REVIEW = "review"
    APPROVED = "approved"
    IMPLEMENTED = "implemented"
    CLOSED = "closed"
    CANCELLED = "cancelled"


class A3SectionType(str, Enum):
    """Standard A3 sections."""
    
    # Problem Solving A3
    BACKGROUND = "background"
    CURRENT_CONDITION = "current_condition"
    GOAL = "goal"
    ROOT_CAUSE = "root_cause"
    COUNTERMEASURES = "countermeasures"
    IMPLEMENTATION_PLAN = "implementation_plan"
    FOLLOW_UP = "follow_up"
    
    # Proposal A3
    PROBLEM_STATEMENT = "problem_statement"
    ANALYSIS = "analysis"
    PROPOSED_SOLUTION = "proposed_solution"
    COST_BENEFIT = "cost_benefit"
    TIMELINE = "timeline"
    RISKS = "risks"
    
    # Custom
    CUSTOM = "custom"


class A3(Base, TimestampMixin, AuditMixin, SoftDeleteMixin):
    """
    A3 problem-solving document.
    
    The A3 is a structured problem-solving approach used in Lean/TPS.
    Named after the A3 paper size (11x17 inches) traditionally used.
    """
    
    __tablename__ = "a3s"
    
    # Identification
    a3_number: Mapped[str] = mapped_column(
        String(50),
        unique=True,
        nullable=False,
        index=True,
    )
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    
    # Type and Status
    a3_type: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        default=A3Type.PROBLEM_SOLVING.value,
        index=True,
    )
    status: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        default=A3Status.DRAFT.value,
        index=True,
    )
    
    # Related Entity (polymorphic reference)
    related_entity_type: Mapped[str | None] = mapped_column(String(50), nullable=True)
    related_entity_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        nullable=True,
        index=True,
    )
    
    # Ownership
    author_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    sponsor_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    coach_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    
    # Team Members
    team_members: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    # List of user IDs
    
    # Dates
    started_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    target_completion_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
        index=True,
    )
    actual_completion_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    
    # Review and Approval
    last_review_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    review_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    approved_by_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    approved_date: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    
    # Progress
    progress_percentage: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    
    # Version Control
    version: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    
    # PDF storage
    pdf_storage_key: Mapped[str | None] = mapped_column(String(500), nullable=True)
    
    # Tags
    tags: Mapped[list | None] = mapped_column(JSONB, nullable=True, default=list)
    
    # Department/Area
    department: Mapped[str | None] = mapped_column(String(100), nullable=True, index=True)
    area: Mapped[str | None] = mapped_column(String(100), nullable=True)
    
    # Priority
    priority: Mapped[str] = mapped_column(String(20), default="medium", nullable=False)
    
    # Summary (for quick view)
    summary: Mapped[str | None] = mapped_column(Text, nullable=True)
    
    # Lessons Learned
    lessons_learned: Mapped[str | None] = mapped_column(Text, nullable=True)
    
    # Yokoten (horizontal deployment) tracking
    is_yokoten_candidate: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
        nullable=False,
    )
    yokoten_areas: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    
    # Custom Fields
    custom_fields: Mapped[dict | None] = mapped_column(JSONB, nullable=True, default=dict)
    
    # Relationships
    author: Mapped["User | None"] = relationship("User", foreign_keys=[author_id])
    sponsor: Mapped["User | None"] = relationship("User", foreign_keys=[sponsor_id])
    coach: Mapped["User | None"] = relationship("User", foreign_keys=[coach_id])
    approved_by: Mapped["User | None"] = relationship("User", foreign_keys=[approved_by_id])
    
    sections: Mapped[list["A3Section"]] = relationship(
        "A3Section",
        back_populates="a3",
        cascade="all, delete-orphan",
        order_by="A3Section.section_order",
        lazy="selectin",
    )
    
    __table_args__ = (
        Index("ix_a3s_type_status", a3_type, status),
        Index("ix_a3s_author_status", author_id, status),
        Index("ix_a3s_department", department),
        Index(
            "ix_a3s_open",
            status,
            postgresql_where=(status.notin_(["closed", "cancelled"])),
        ),
    )
    
    def update_progress(self) -> None:
        """Calculate progress based on section completion."""
        if not self.sections:
            self.progress_percentage = 0
            return
        
        total_sections = len(self.sections)
        completed_sections = sum(1 for s in self.sections if s.is_complete)
        
        self.progress_percentage = int((completed_sections / total_sections) * 100)
    
    @property
    def is_open(self) -> bool:
        """Check if A3 is still open."""
        return self.status not in [A3Status.CLOSED.value, A3Status.CANCELLED.value]
    
    @property
    def is_overdue(self) -> bool:
        """Check if A3 is overdue."""
        if self.target_completion_date is None:
            return False
        if not self.is_open:
            return False
        from datetime import timezone as tz
        return datetime.now(tz.utc) > self.target_completion_date


class A3Section(Base, TimestampMixin):
    """
    Individual section of an A3 document.
    
    Each A3 consists of multiple sections following a structured template.
    """
    
    __tablename__ = "a3_sections"
    
    a3_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("a3s.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    
    # Section Identification
    section_type: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        default=A3SectionType.CUSTOM.value,
    )
    section_name: Mapped[str] = mapped_column(String(100), nullable=False)
    section_order: Mapped[int] = mapped_column(Integer, nullable=False)
    
    # Content
    content: Mapped[str | None] = mapped_column(Text, nullable=True)
    
    # Structured Content (for forms/diagrams)
    structured_content: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    # Can contain:
    # - 5 Whys analysis
    # - Fishbone diagram data
    # - Action items
    # - Metrics/KPIs
    # - Gantt chart data
    
    # Completion
    is_complete: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )
    completed_by_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    
    # Guidance text for this section
    guidance: Mapped[str | None] = mapped_column(Text, nullable=True)
    
    # Attachments
    attachments: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    
    # Comments/Feedback
    comments: Mapped[list | None] = mapped_column(JSONB, nullable=True)
    
    # Version
    version: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    
    # Relationships
    a3: Mapped["A3"] = relationship("A3", back_populates="sections")
    completed_by: Mapped["User | None"] = relationship("User")
    
    __table_args__ = (
        UniqueConstraint("a3_id", "section_order", name="uq_a3_section_order"),
        Index("ix_a3_sections_a3_order", a3_id, section_order),
    )


# Default section templates for different A3 types
A3_SECTION_TEMPLATES = {
    A3Type.PROBLEM_SOLVING.value: [
        {
            "type": A3SectionType.BACKGROUND.value,
            "name": "Background",
            "guidance": "Provide context. Why is this problem important? What business need does it address?",
        },
        {
            "type": A3SectionType.CURRENT_CONDITION.value,
            "name": "Current Condition",
            "guidance": "Describe the current state. Use data and facts. What is actually happening?",
        },
        {
            "type": A3SectionType.GOAL.value,
            "name": "Goal/Target Condition",
            "guidance": "What specific, measurable outcome do you want to achieve? By when?",
        },
        {
            "type": A3SectionType.ROOT_CAUSE.value,
            "name": "Root Cause Analysis",
            "guidance": "Use 5 Whys, Fishbone, or other tools. What is the real cause of the problem?",
        },
        {
            "type": A3SectionType.COUNTERMEASURES.value,
            "name": "Countermeasures",
            "guidance": "What specific actions will address the root cause? Who, what, when?",
        },
        {
            "type": A3SectionType.IMPLEMENTATION_PLAN.value,
            "name": "Implementation Plan",
            "guidance": "Timeline, resources, responsibilities. How will you implement?",
        },
        {
            "type": A3SectionType.FOLLOW_UP.value,
            "name": "Follow-Up",
            "guidance": "How will you verify results? What metrics will you track?",
        },
    ],
    A3Type.PROPOSAL.value: [
        {
            "type": A3SectionType.BACKGROUND.value,
            "name": "Background",
            "guidance": "Context and business case for the proposal.",
        },
        {
            "type": A3SectionType.PROBLEM_STATEMENT.value,
            "name": "Problem/Opportunity",
            "guidance": "What problem are you solving or opportunity are you addressing?",
        },
        {
            "type": A3SectionType.ANALYSIS.value,
            "name": "Analysis",
            "guidance": "Data and analysis supporting your proposal.",
        },
        {
            "type": A3SectionType.PROPOSED_SOLUTION.value,
            "name": "Proposed Solution",
            "guidance": "Your recommended approach. Why this solution?",
        },
        {
            "type": A3SectionType.COST_BENEFIT.value,
            "name": "Cost/Benefit Analysis",
            "guidance": "Investment required and expected return.",
        },
        {
            "type": A3SectionType.TIMELINE.value,
            "name": "Timeline",
            "guidance": "Implementation schedule and milestones.",
        },
        {
            "type": A3SectionType.RISKS.value,
            "name": "Risks and Mitigation",
            "guidance": "Potential risks and how you will address them.",
        },
    ],
}

File: frontend/src/stores/a3.ts
import { create } from 'zustand';
import { devtools, persist } from 'zustand/middleware';
import { apiClient } from '@/api/client';

type A3Type = 'problem_solving' | 'proposal' | 'status_report' | 'strategy';
type A3Status = 'draft' | 'in_progress' | 'review' | 'approved' | 'implemented' | 'closed' | 'cancelled';
type A3Priority = 'critical' | 'high' | 'medium' | 'low';
type A3SectionType = 
  | 'background' | 'current_condition' | 'goal' | 'root_cause' 
  | 'countermeasures' | 'implementation_plan' | 'follow_up'
  | 'problem_statement' | 'analysis' | 'proposed_solution'
  | 'cost_benefit' | 'timeline' | 'risks' | 'custom';

interface A3Section {
  id: string;
  a3_id: string;
  section_type: A3SectionType;
  section_name: string;
  section_order: number;
  content?: string;
  structured_content?: any;
  is_complete: boolean;
  completed_at?: string;
  completed_by_id?: string;
  completed_by_name?: string;
  guidance?: string;
  attachments?: any[];
  comments?: any[];
  version: number;
  created_at: string;
  updated_at: string;
}

interface A3 {
  id: string;
  a3_number: string;
  title: string;
  a3_type: A3Type;
  status: A3Status;
  related_entity_type?: string;
  related_entity_id?: string;
  author_id: string;
  author_name: string;
  sponsor_id?: string;
  sponsor_name?: string;
  coach_id?: string;
  coach_name?: string;
  team_members?: string[];
  started_date?: string;
  target_completion_date?: string;
  actual_completion_date?: string;
  last_review_date?: string;
  review_notes?: string;
  approved_by_id?: string;
  approved_by_name?: string;
  approved_date?: string;
  progress_percentage: number;
  version: number;
  pdf_storage_key?: string;
  tags?: string[];
  department?: string;
  area?: string;
  priority: A3Priority;
  summary?: string;
  lessons_learned?: string;
  is_yokoten_candidate: boolean;
  yokoten_areas?: string[];
  custom_fields?: Record<string, any>;
  sections: A3Section[];
  created_at: string;
  updated_at: string;
  created_by_id: string;
  created_by_name: string;
}

interface A3Stats {
  total: number;
  by_status: Record<A3Status, number>;
  by_type: Record<A3Type, number>;
  by_priority: Record<A3Priority, number>;
  overdue_count: number;
  avg_completion_days: number;
  approval_pending: number;
}

interface A3State {
  a3s: A3[];
  stats: A3Stats;
  isLoading: boolean;
  error: string | null;
  lastFetchedAt: number | null;

  fetchA3s: () => Promise<void>;
  fetchA3ById: (id: string) => Promise<A3 | null>;
  createA3: (a3: Partial<A3>) => Promise<A3>;
  updateA3: (id: string, updates: Partial<A3>) => Promise<A3>;
  deleteA3: (id: string) => Promise<void>;
  updateSection: (a3Id: string, sectionId: string, updates: Partial<A3Section>) => Promise<A3Section>;
  completeSection: (a3Id: string, sectionId: string) => Promise<void>;
  submitForReview: (id: string) => Promise<void>;
  approve: (id: string, notes?: string) => Promise<void>;
  reject: (id: string, notes: string) => Promise<void>;
  exportPDF: (id: string) => Promise<void>;
  clearError: () => void;
}

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

const initialStats: A3Stats = {
  total: 0,
  by_status: {
    draft: 0,
    in_progress: 0,
    review: 0,
    approved: 0,
    implemented: 0,
    closed: 0,
    cancelled: 0,
  },
  by_type: {
    problem_solving: 0,
    proposal: 0,
    status_report: 0,
    strategy: 0,
  },
  by_priority: {
    critical: 0,
    high: 0,
    medium: 0,
    low: 0,
  },
  overdue_count: 0,
  avg_completion_days: 0,
  approval_pending: 0,
};

export const useA3Store = create<A3State>()(
  devtools(
    persist(
      (set, get) => ({
        a3s: [],
        stats: initialStats,
        isLoading: false,
        error: null,
        lastFetchedAt: null,

        fetchA3s: async () => {
          const { lastFetchedAt, isLoading } = get();
          const now = Date.now();

          if (isLoading) {
            return;
          }

          // Cache for 30 seconds
          if (lastFetchedAt && now - lastFetchedAt < 30000) {
            return;
          }

          set({ isLoading: true, error: null });
          try {
            const data = await apiClient.get<any>('/a3s');
            const a3s: A3[] = data.items || [];

            // Calculate stats
            const stats: A3Stats = {
              total: a3s.length,
              by_status: a3s.reduce((acc, a3) => {
                acc[a3.status] = (acc[a3.status] || 0) + 1;
                return acc;
              }, {} as Record<A3Status, number>),
              by_type: a3s.reduce((acc, a3) => {
                acc[a3.a3_type] = (acc[a3.a3_type] || 0) + 1;
                return acc;
              }, {} as Record<A3Type, number>),
              by_priority: a3s.reduce((acc, a3) => {
                acc[a3.priority] = (acc[a3.priority] || 0) + 1;
                return acc;
              }, {} as Record<A3Priority, number>),
              overdue_count: a3s.filter(a3 => {
                if (!a3.target_completion_date) return false;
                if (['closed', 'cancelled'].includes(a3.status)) return false;
                return new Date(a3.target_completion_date) < new Date();
              }).length,
              avg_completion_days: 0, // Calculated on backend
              approval_pending: a3s.filter(a3 => a3.status === 'review').length,
            };

            set({
              a3s,
              stats,
              lastFetchedAt: now,
              isLoading: false,
            });
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch A3s',
              isLoading: false,
            });
          }
        },

        fetchA3ById: async (id: string) => {
          set({ isLoading: true, error: null });
          try {
            const a3 = await apiClient.get<A3>(`/a3s/${id}`);

            set(state => ({
              a3s: state.a3s.map(a => a.id === id ? a3 : a),
              isLoading: false,
            }));

            return a3;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch A3',
              isLoading: false,
            });
            return null;
          }
        },

        createA3: async (a3Data: Partial<A3>) => {
          set({ isLoading: true, error: null });
          try {
            const a3 = await apiClient.post<A3>('/a3s', a3Data);

            set(state => ({
              a3s: [a3, ...state.a3s],
              stats: {
                ...state.stats,
                total: state.stats.total + 1,
              },
              isLoading: false,
            }));

            return a3;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to create A3',
              isLoading: false,
            });
            throw error;
          }
        },

        updateA3: async (id: string, updates: Partial<A3>) => {
          set({ isLoading: true, error: null });
          try {
            const a3 = await apiClient.patch<A3>(`/a3s/${id}`, updates);

            set(state => ({
              a3s: state.a3s.map(a => a.id === id ? a3 : a),
              isLoading: false,
            }));

            return a3;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to update A3',
              isLoading: false,
            });
            throw error;
          }
        },

        deleteA3: async (id: string) => {
          set({ isLoading: true, error: null });
          try {
            await apiClient.delete(`/a3s/${id}`);

            set(state => ({
              a3s: state.a3s.filter(a => a.id !== id),
              stats: {
                ...state.stats,
                total: Math.max(0, state.stats.total - 1),
              },
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to delete A3',
              isLoading: false,
            });
            throw error;
          }
        },

        updateSection: async (a3Id: string, sectionId: string, updates: Partial<A3Section>) => {
          set({ isLoading: true, error: null });
          try {
            const section = await apiClient.patch<A3Section>(`/a3s/${a3Id}/sections/${sectionId}`, updates);

            // Refresh A3 to get updated progress
            await get().fetchA3ById(a3Id);

            set({ isLoading: false });
            return section;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to update section',
              isLoading: false,
            });
            throw error;
          }
        },

        completeSection: async (a3Id: string, sectionId: string) => {
          await get().updateSection(a3Id, sectionId, { is_complete: true });
        },

        submitForReview: async (id: string) => {
          set({ isLoading: true, error: null });
          try {
            const a3 = await apiClient.post<A3>(`/a3s/${id}/submit`);

            set(state => ({
              a3s: state.a3s.map(a => a.id === id ? a3 : a),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to submit A3',
              isLoading: false,
            });
            throw error;
          }
        },

        approve: async (id: string, notes?: string) => {
          set({ isLoading: true, error: null });
          try {
            const a3 = await apiClient.post<A3>(`/a3s/${id}/approve`, { notes });

            set(state => ({
              a3s: state.a3s.map(a => a.id === id ? a3 : a),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to approve A3',
              isLoading: false,
            });
            throw error;
          }
        },

        reject: async (id: string, notes: string) => {
          set({ isLoading: true, error: null });
          try {
            const a3 = await apiClient.post<A3>(`/a3s/${id}/reject`, { notes });

            set(state => ({
              a3s: state.a3s.map(a => a.id === id ? a3 : a),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to reject A3',
              isLoading: false,
            });
            throw error;
          }
        },

        exportPDF: async (id: string) => {
          set({ isLoading: true, error: null });
          try {
            const response = await fetch(`${API_BASE_URL}/a3s/${id}/export`, {
              headers: {
                'Authorization': `Bearer ${localStorage.getItem('access_token')}`,
              },
            });

            if (!response.ok) {
              throw new Error(`Failed to export A3: ${response.statusText}`);
            }

            const blob = await response.blob();
            const url = window.URL.createObjectURL(blob);
            const a = document.createElement('a');
            a.href = url;
            a.download = `a3_${id}_${new Date().toISOString().split('T')[0]}.pdf`;
            document.body.appendChild(a);
            a.click();
            window.URL.revokeObjectURL(url);
            document.body.removeChild(a);

            set({ isLoading: false });
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to export A3',
              isLoading: false,
            });
            throw error;
          }
        },

        clearError: () => set({ error: null }),
      }),
      {
        name: 'a3-storage',
        partialize: (state) => ({
          a3s: state.a3s,
          stats: state.stats,
          lastFetchedAt: state.lastFetchedAt,
        }),
      }
    )
  )
);
